consensus/istanbul/backend/backend.go

```diff
@@ -848,7 +851,66 @@ func (sb *Backend) retrieveRegisteredAndElectedValidators() (map[common.Address]
 	}
 
 	// Retrieve the registered/elected valset from the smart contract (for the registered validators) and headers (for the elected validators).
+	validatorsSet, err := sb.retrieveUncachedRegisteredAndElectedValidators()
+	if err != nil {
+		return nil, err
+	}
+
+	sb.cachedRegisteredElectedValSet = validatorsSet
+	sb.cachedRegisteredElectedValSetTimestamp = time.Now()
+
+	return validatorsSet, nil
+}
 
+// retrieveMaybeStaleRegisteredAndElectedValidators returns the most recently
+// cached set of registered and elected validators and asynchronously updates
+// the cache if it is older than 1 minute. If no set has ever been cached, this
+// function may block for a couple seconds while retrieving the uncached set.
+func (sb *Backend) retrieveMaybeStaleRegisteredAndElectedValidators() (map[common.Address]bool, error) {
+	sb.cachedRegisteredElectedValSetMu.Lock()
```

Comment:
Changed it to a RWMutex that doesn't lock when calling the function that will take a while

Reply:
One result of this implementation, is that if a barrage of connection requests come in together when the cache needs to be updated, it will kick off a goroutine to update it for each one, until it is finally updated. It would be best to have machoism to ensure only one task is created to update the cache.

Reply:
Great point-- I added some logic to make sure that only one goroutine is doing the work of `updateCachedValidatorConnSet` at a time. If another goroutine is currently doing the work of updating the cache, the new goroutine will wait for the work to be finished & return any error that arose in the work-goroutine

It does this using `sync.Cond`, here's a playground I made that does something similar: https://play.golang.org/p/MvpS1Zm-pnw

If you have a different solution in mind maybe this should be left to a subsequent PR?